Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,167 @@
+
+[User-space memory allocations](User-space-memory-allocations)
+
+[License: CC BY-SA 4.0](#License-CC-BY-SA-40)
+
+
+## **Terms and Abbreviations**
+Plese refer to the Linux Kernel documentation.
+   
+
+## **References**
+
+1. [Linux Kernel website](https://www.kernel.org) - <https://www.kernel.org>
+2. ***Interference Scenarios for an ARM64 Linux System***
+3. [CC BY-SA 4.0 Deed | Attribution-ShareAlike 4.0 International | Creative Commons](https://creativecommons.org/licenses/by-sa/4.0/) - <https://creativecommons.org/licenses/by-sa/4.0/> License
+
+
+## **Disclaimer**
+* This document is not intended to be a replacement for understanding the memory management of the Linux,
+  nor it attempts to be an exhaustive analysis of safety implications.
+* Because of the very volatile nature of the code within the Linux Kernel, each of the statements made
+  below should not be taken at face value, but rather verified, for any Linux kernel version following
+  the one used while writing the document.
+* When referring to specific HW features, the document refers to the ARM64 architecture.
+
+## **Purpose of the document**
+This document aims to providing an holistic view of what happens in Linux memory management, so that
+one is at least aware of certain features and can use this document as jumping pad toward more detailed documentation.
+Or even to the code base.
+
+## **Structure of the document**
+The document is divided in two parts, based on hte destinatary of the memory allocations discussed: kernel-space or user-space.
+Individual points are numbered for ease of reference, but the numbering is not meant to represent any sequence.
+
+## **Memory management in Linux**
+
+### **Kernel-space memory allocations**
+
+#### **Facts**
+1. unlike processes memory, kernel memory pages are not swapped, nor dropped silently by the kernel itself,
+   although an hypervisor will do to a VM what the kernel does to a process (but this is beyond the control of the kernel)
+2. the kernel context (usually EL1 on ARM64) uses one single memory map (page tables) across all the cores
+   executing in kernel mode
+3. on 64 bit systems (e.g. ARM64 and x86_64), usually almost all physical memory is mapped in a
+   (semi)contiguous (there can be holes) range. Memory within this range is both virtually and physically contiguous.
+4. physically contiguous memory is treated as a scarce resource, and typically is not provided to userspace,
+   unless it explicitly asks for it (e.g. for implementing some DMA buffer)
+5. the kernel can access userspace in two ways:
+   1. through the userspace mapping
+      1. it is done in exceptional cases, like copy_to_user_copy_from_user
+      2. it uses the userspace memory map, which can implement HW protections against kernel misuses
+      3. it allows the kernel to see process pages in the same sequence and with the same mappings as the proces does
+   2. through the EL1 mappings + linear map
```

Comment:
* ok on the style 
* the security aspect is there because I'm afraid that if I do not explain it upfront, it will be periodically brought up, since I noticed that people tend to confuse the intent behind certain security features and think that said features provide more protection that they actually do
* the kernel doesn't intrinsically "know" if a page is in use by user space or by itself, because in both cases it is always mapped. The problem is that a broken driver or a broken kernel component can cause cascaded effects, so it's not so straightforward to say where the wrong access cancome from. It might come from a legitimate and "sane" component that has had its metadata corrupted by another one. I can elaborate more on this.

Reply:
The rewording should now address the comments above.